src/openms_gui/source/VISUAL/TVToolDiscovery.cpp

```diff
@@ -55,87 +57,182 @@ namespace OpenMS
+    plugin_params_.clear();
+    waitForPluginParams();
+    return plugin_params_;
+  }
 
+  Param TVToolDiscovery::getParamFromIni_(const String &tool_path, bool plugins)
+  {
+    FileHandler fh;
     // Temporary file path and arguments
     String path = File::getTemporaryFile();
     String working_dir = path.prefix(path.find_last_of('/'));
     QStringList args{"-write_ini", path.toQString()};
-    // Start tool/util to write the ini file
     Param tool_param;
     String executable;
     // Return empty param if tool executable cannot be found
     try
     {
-      executable = File::findSiblingTOPPExecutable(tool_name);
+      // Is an executable already or has a sibling Executable
+      executable = File::exists(tool_path) ? tool_path : File::findSiblingTOPPExecutable(tool_path);
     }
     catch (const Exception::FileNotFound& e)
     {
-      std::cerr << "TOPP tool: " << e << " not found during tool discovery. Skipping." << std::endl;
+      OPENMS_LOG_WARN << "TOPP tool: " << e << " not found during tool discovery. Skipping." << std::endl;
       return tool_param;
     }
-    // Spawning a thread for all tools is no problem (if std::async decides to do so)
-    // but spawning that many processes failed with not enough file handles on machines with large number of cores.
-    // Restricting the number of running processes solves that issue.
-    while (running_processes >= 6) { std::this_thread::sleep_for(std::chrono::milliseconds(10)); QCoreApplication::processEvents(); }
-    ++running_processes;
+
     // Write tool ini to temporary file
-    auto return_state = ExternalProcess().run(executable.toQString(), args, working_dir.toQString(), false, ExternalProcess::IO_MODE::NO_IO);
-    --running_processes;
+    auto lam_out = [&](const String& out) { OPENMS_LOG_INFO << out; };
+    auto lam_err = [&](const String& out) { OPENMS_LOG_INFO << out; };
+    ExternalProcess proc(lam_out, lam_err);
+    auto return_state = proc.run(executable.toQString(), args, working_dir.toQString(), true, ExternalProcess::IO_MODE::NO_IO);
     // Return empty param if writing the ini file failed
     if (return_state != ExternalProcess::RETURNSTATE::SUCCESS)
     {
+      OPENMS_LOG_WARN << "TOPP tool: " << executable << " error during execution: " << (uint32_t)return_state << "\n";
       return tool_param;
     }
     // Parse ini file to param object
     ParamXMLFile paramFile;
-    paramFile.load((path).c_str(), tool_param);
+    try
+    {
+      paramFile.load((path).c_str(), tool_param);
+    }
+    catch(const Exception::FileNotFound& e)
+    {
+      OPENMS_LOG_WARN << e << "\n" << "TOPP tool: " << executable <<
+        " not able to write ini. Plugins must implement -write_ini parameter. Skipping." << std::endl;
+      return tool_param;
+    }
+    
+    if (plugins)
+    {
+      auto tool_name = tool_param.begin().getTrace().begin()->name;
+      auto filename = tool_path.substr(tool_path.find_first_of('/') + 1);
```

Comment:
is this working as intended?
searching for the first `/` could still be part of a larger path? I'm not sure what the content of tool_path is, but it seems its obtained from `getPlugins_()` and this could be anything?

Reply:
why not use `std::filesystem::path(tool_path).filename()` or similar?